### ISSUE TITLE
How to Create and Delete References Between Entities

### ISSUE THREAD

**Issue by michaelgr1** [OData/AspNetCoreOData#132]
How do I post and put related entities. I have a Product entity for example with a Supplier navigation property. I would like to have an endpoint that allows to change the supplier for an existing product. Using odata I was able to make a post endpoint for the Product, but if I tried to include the supplier it had to be a new supplier, an existing one would through an error regarding the same primary key already exists. I am trying to allow posting an existing supplier and only binding it, without trying to create it.
I found a tutorial on microsoft.com, where they showed how to use CreateRef and DeleteRef. I tried to implement the same thing, but the endpoint wasn't even created. I assume the tutorial was for an earlier version, I am using odata 8.0.0-rc. I tried cloning the repository and running the sample project. But the above methods seem to not really be implemented (they return a string), and the endpoint is not mapped anyhow.
So, I am looking for an example on how to implement the above behavior. Any help would be appreciated, and I am sorry if I missed something obvious.
Thanks, Michael.

**Comment by dxrdxr** [comment]
I have a similar situation, I have a Diver that contains a reference to a Contact that usually exists. In the simple case where the Contact doesn't exist I POST the following to do a Deep Insert (note no keys):
```
{
    "LastDive": "2099-01-01T00:00:00Z",
    "LoggedDives": 9999,
    "Contact": {
        "FirstName": "Delete",
        "MiddleName": "Test",
        "LastName": "Item",
    }
}
```

In the common case, the Contact already exists so I don't want to create a new one, just reuse the existing one. So I bind the two:
```
{
    "LastDive": "2099-01-01T00:00:00Z",
    "LoggedDives": 9999,
    "Contact@odata.bind": "/Contacts(123456)"
}
```
In the DiversController `Post([FromBody] Diver diver)` I have to detect these different cases. In the second case OData will be passing in a full `Diver` but the navigation property `Diver.Contact` will only have the key `ID` filled in to be "123456". Depending on your backing database, you either retrieve the Contact or insert it by reference. 

In my implementation I use EF Core so I `Attach` the "stub" Contact with just the ID, `Add` the Diver, and `SaveChanges`:

```
    diverContext.Contacts.Attach(diver.Contact);
    diverContext.Divers.Add(diver);
    diverContext.SaveChanges();
```

**Comment by michaelgr1** [comment]
@dxrdxr I tried implementing that approach. But then, if I want to post the related entity I need to include all the information to make the model state valid. It seems a bit weird to send all the information when I am actually only interested in some of it. Maybe I can do the same thing with patch? Anyhow, thanks for the suggestions

**Comment by michaelgr1** [comment]
To solve the problem I created the following DTO:

    public class BindDto<T> {

        [Required]
        public T Id { get; set; }

    }

Then, I created an action inside my product controller:

        [HttpPost]
        public async Task<IActionResult> BindSupplier([FromODataUri] int key, [FromBody] BindDto<int> bind) {
            if (!ModelState.IsValid) {
                return BadRequest(ModelState);
            }

            var product = await this.db.Products.FindAsync(key);

            var supplier = await this.db.Suppliers.FindAsync(bind.Id);

            if (product == null || supplier == null) {
                return NotFound();
            }

            product.Supplier = supplier;
            await this.db.SaveChangesAsync();
            return NoContent();
        }


I added this method as an action bound to the product entity using the edm builder:

`builder.EntityType<Product>().Action("BindSupplier");`

I don't know if that is the right way to do it, but it works for me and seems more logical the sending my entire product class when I am only interested in changing the supplier.
But if you know how am I supposed to do it, please let me know.

**Comment by dxrdxr** [comment]
> @dxrdxr I tried implementing that approach. But then, if I want to post the related entity I need to include all the information to make the model state valid. It seems a bit weird to send all the information when I am actually only interested in some of it. Maybe I can do the same thing with patch? Anyhow, thanks for the suggestions

Yes, in my case I had to `ModelState.Remove` the Required properties to get it to work. Patch doesn't work any better than Put. I agree that it is weird that the validation code is running when you are explicitly trying to just Bind a reference.

**Comment by julealgon** [comment]
> Using a trivial BindDto looks to work too, but both of these ways just doesn't seem "right".

You are correct, they are not @dxrdxr . The OData spec supports posting and updating nested data and nested relations, but this is not trivial to implement as it is not automatic.

@michaelgr1 it seems to me that your first problem is trying to use `POST` for this. You see, `POST` is to be used for creating an entity, not updating it. Since you are attempting to update an existing `Product` entity, indeed it doesn't make sense to use `POST` for that (and thus why it feels clunky having to POST the entire entity again).

For updating partially like this, your best bet would be a `PATCH` endpoint instead. Instead of `POST /api/products`, you do a `PATCH /api/products/{productId}`. The real challenge though is how to inform the server that you just want to add a relationship. You could just send the `Suppliers` property and have only the `Id` field populated, and then on the server handle this special case by explicitly checking for an ID only supplier and adding the relationship. As far as I'm concerned, there is no automatic way of doing this (using a standard `Delta<Product>` for instance): you have to manually check the resulting state of the model to see that the rest of the data for the supplier is missing but that it has a valid ID, and only then make the association.

I posted a question on StackOverflow some time ago that is quite a bit similar to the problem you are describing:
- https://stackoverflow.com/questions/61229762/how-to-post-odata-entity-and-link-it-to-multiple-existing-entities-at-the-same-t

You can see that the answer is quite detailed there. Chris' answer indicates that one should not update related entities on a PATCH request, but I'm not sure that's actually really the case: I don't personally remember seeing that restriction in the spec. According to him, the cleanest approach would be to have separate endpoints, and have a dedicated one for updating the reference (this is natively supported in OData). Batch can then be used to combine these into a single request if needed (for example, POST product, followed by a second POST/PUT for the reference).

You can see more details about manipulating links here:
- https://docs.microsoft.com/en-us/odata/webapi/built-in-routing-conventions

Check the **Querying, Creating and Deleting Links** section (this is the ref approach mentioned above)

Lastly, [the `JsonPatch` spec](https://docs.microsoft.com/en-us/aspnet/core/web-api/jsonpatch?view=aspnetcore-6.0) also defines a way to include nested entities, but it seems to only support the entity itself, and not references to existing entities. I'm not quite sure however about the integration of `JsonPatch` with OData.
